mutable_tree.go

```diff
@@ -759,12 +760,46 @@ func (tree *MutableTree) SaveVersion() ([]byte, int64, error) {
 	tree.ImmutableTree = tree.ImmutableTree.clone()
 	tree.lastSaved = tree.ImmutableTree.clone()
 	tree.orphans = map[string]int64{}
-	tree.unsavedFastNodeAdditions = make(map[string]*FastNode)
+	tree.unsavedFastNodeAdditions = make(map[string]*types.FastNode)
 	tree.unsavedFastNodeRemovals = make(map[string]interface{})
 
 	return tree.Hash(), version, nil
 }
 
+// SaveBranch recursively saves all updated descendants of a node.
+// then it computes its hash.
+// A descendant is tracked for whether or not its updated, based on if
+// node.persisted is true.
+// TODO: Make better static guarantee in tree,
+// for what nodes exist in deserialized memory, and just not getting extra nodes
+// and a clear decision criteria between "no branch node", "new branch with as of yet unknown hash"
+func (tree *MutableTree) saveBranch(node *Node) []byte {
+	if node.persisted {
+		return node.hash
+	}
+
+	if node.leftNode != nil {
+		node.leftHash = tree.saveBranch(node.leftNode)
+	}
+	if node.rightNode != nil {
+		node.rightHash = tree.saveBranch(node.rightNode)
+	}
+
+	node._hash()
+	tree.ndb.SaveNode(node)
+
+	// resetBatch only working on generate a genesis block
+	// TODO: What is this? It should almost certainly be deleted / moved elsewhere.
+	if node.version <= genesisVersion {
+		tree.ndb.resetBatch()
+	}
+	// TODO: This should be deletable? Is it just here for garbage collection purposes?
```

Comment:
I think you are right. We should probably run benchmarks before removing them. Wdyt?